biz.aQute.bnd.facade.provider/src/biz/aQute/bnd/facade/provider/FacadeManagerProvider.java

```diff
@@ -0,0 +1,212 @@
+package biz.aQute.bnd.facade.provider;
+
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.Constants;
+import org.osgi.framework.Filter;
+import org.osgi.framework.InvalidSyntaxException;
+import org.osgi.framework.ServiceObjects;
+import org.osgi.framework.ServiceReference;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.util.tracker.ServiceTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import biz.aQute.bnd.facade.api.Binder;
+import biz.aQute.bnd.facade.api.Delegate;
+import biz.aQute.bnd.facade.api.FacadeManager;
+import biz.aQute.bnd.facade.api.Instance;
+
+@Component
+@SuppressWarnings({ "rawtypes", "unchecked" })
```

Comment:
This code is about getting the types correct dynamically. Since you do not get a type from the service registry and the map holding the objects cannot distinguish the object generic type you end up with just <?> and then in many cases you need to cast and jump through hoops due to limitations of the generic type system.

Generics are fantastic for the API, but in this class you just do not get any advantage using generics. It is all manual checking. I found over time that you either get 30% more horrible casts and silly conversion or easy to read code when you leave out the generics.